Added error check to prevent segfault

DanielYankura opened the pull request:
<!--
If this PR is associated with an issue be sure to reference it
by including "refs #<issue>" or "closes #<issue>" (e.g., #closes #1234).

If this PR implements an enhancement that does not have an existing issue, please add the following:
-->

## Reason
<!--Why do you need this feature or what is the enhancement?-->
- Users may not know why it's segfaulting
## Design
<!--A concise description (design) of the enhancement.-->
- If --check-input is run with any param that creates a final task it results in a segfault
- Added check right after final tasks are set that checks if --check-input is being used as well
## Impact
<!--Will the enhancement change existing APIs or add something new?-->
Will aid in debugging MOOSE inputs

<!--
If this PR implements a bug fix, please create an issue for the bug and reference the issue.
-->
closes #29788

DanielYankura committed:
Added error check to prevent segfault

- If --check-input is run with any param that creates a final task it results in a segfault
- Added check right after final tasks are set that checks if --check-input is being used as well

closes #29788

pbehne reviewed (commented):
While throwing an error is certainly more desirable than allowing a segfault, it would be more useful if we could identify the cause of the segfault and address that directly. Have you looked into this?

DanielYankura commented:
I looked into that. When using --split-mesh or --mesh-only it sets those as the final task. --check-input isn't a task so it still get's used and calls MooseApp::errorCheck() which tries calling feProblem(). Because not all tasks ran, calling feProblem() results in a segfault.

pbehne commented:
> I looked into that. When using --split-mesh or --mesh-only it sets those as the final task. --check-input isn't a task so it still get's used and calls MooseApp::errorCheck() which tries calling feProblem(). Because not all tasks ran, calling feProblem() results in a segfault.

Okay, makes sense. To reiterate, the documentation for the `--check-input` option is:

> Check the input file (i.e. requires -i <filename>) and quit

With this in mind, it doesn't make sense to call with `--mesh-only` in the first place. Like you pointed out about some tasks not running, some things aren't getting set up, which causes issues.

pbehne commented on framework/include/actions/ActionWarehouse.h:
```diff
@@ -297,6 +297,8 @@ class ActionWarehouse : public ConsoleStreamInterface
    */
   bool isTaskComplete(const std::string & task) const;
 
+  std::string getFinalTask() { return _final_task; }
```
```suggestion
  std::string & getFinalTask() const { return _final_task; }
```

pbehne replied:
This is no longer used, and thus no longer needed, correct?

pbehne reviewed (changes requested):
When I run moose with options `--check-input` and `--version`, I get a segfault. Could you look into this and some of the other options to make sure our approach is general enough to catch these?

shikhar413 commented:
Would it make sense to check if feProblem() is not null and error out if this is null and _check_inputs is true? This check can be applied right before calling `errorCheck()` in MooseApp.C

This should make the error check more robust regardless of what option is specified

DanielYankura commented:
I did some more digging and I think I found a way to make it more general. If --check-input is used with any parameter that sets _ready_to_exit=true it will result in a segfault. I'll move the test and rewrite it so that it catches all of these cases.

DanielYankura commented:
@shikhar413 I just saw your comment. I like that idea. The only issue might be that it'd be hard to give a helpful error message if we don't know why it's null?

shikhar413 commented:
Yeah it might be hard to do that without hard coding which options don’t work with check-inputs. I would say it’s probably fine to just say something like “Command line options are not compatible with —check-inputs as this option requires an FEProblem to be defined in the input file” but someone else might have a different opinion.

The reason that restructuring that logic to prevent ready_to_exit to not be mixed with check-inputs may not be the ideal path forward is that we might in the future want to support mesh-only with check-inputs with a dummy FEProblem instance, and so there’d have to be an exception to that logic if we did that.

shikhar413 commented:
I’ll let someone else make the final call on this, I don’t particularly feel strongly about having the logic be a certain way

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/2667492/), step Docs: sync website on ce16e33 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/29803/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/2667508/), step Generate coverage on ce16e33 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>4965c6</th>
      <th colspan=4>#29803 ce16e3</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.23%</td>
      <td>85.24%</td>
      <td>+0.00%</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>108491</td>
      <td>108495</td>
      <td>+4</td>
      <td>4</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>18794</td>
      <td>18794</td>
      <td>-</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/29803/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/29803/coverage/framework)


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/29803/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/29803/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/29803/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/29803/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/29803/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/29803/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/29803/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/29803/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/29803/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/29803/coverage/geochemistry)
- [`heat_transfer`](https://mooseframework.inl.gov/docs/PRs/29803/coverage/heat_transfer)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/29803/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/29803/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/29803/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/29803/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/29803/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/29803/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/29803/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/29803/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/29803/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/29803/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/29803/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/29803/coverage/scalar_transport)
- [`solid_mechanics`](https://mooseframework.inl.gov/docs/PRs/29803/coverage/solid_mechanics)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/29803/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/29803/coverage/stochastic_tools)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/29803/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/29803/coverage/xfem)

</details>



This comment will be updated on new commits.

DanielYankura commented:
I feel as well that getting `--check-error` to run with certain other params should be a future issue. The params that cause it to segfault all quit once they're done as well which is why they aren't really compatible. Also keeping track of how many it's interfering with might be difficult because as far as I can tell if there are multiple it will only run one and exit.

I did have a thought for a better way to fix this issue though. If we want to tell the user which command-line argument is interfering I can change `_ready_to_exit` to a string that tells us which param it exited on. That way I can use that info later to know which argument is causing the segfault. This could also come in handy for future issues if we ever need more info for other error checks? I like @shikhar413's idea of checking for a null feproblem instead so I can have it print one error statement if it's null and a different one if it's null and `_ready_to_exit` isn't empty? If that's too much I can just check if it's null and make the error statement more generic.

DanielYankura commented:
Also I just realized it's not `FEProblem` that's not initialized it's `_executor`. The uninitialized `_executor` calling `feProblem()` is what's causing the segfault. Not a huge difference in how to check for errors but I figured it was important to clarify

shikhar413 commented:
> Also I just realized it's not FEProblem that's not initialized it's _executor. The uninitialized _executor calling feProblem() is what's causing the segfault. Not a huge difference in how to check for errors but I figured it was important to clarify

That's good information to know, thanks for digging into this. I'm planning to work on this issue in the near term, and I think some of these fixes will come into play, so I will keep these changes in mind

https://github.com/idaholab/moose/issues/14550

DanielYankura committed:
rewrite of parameter check to prevent segfault

- _ready_to_exit is now a string that contains the parameter name that caused the exit
- Error check was moved to MooseApp::errorCheck()
- Checks fs _ready_to_exit is initialized and _executor/_executione is not initialized
- Throws error that tells the user which parameter is incompatible with --check-input
- Test was updated to match updated error message

pbehne commented on framework/src/base/MooseApp.C:
```diff
@@ -814,13 +814,13 @@ MooseApp::setupOptions()
   {
     Moose::out << getPrintableVersion() << std::endl;
-    _ready_to_exit = true;
+    _ready_to_exit = "display_version";
```
```suggestion
    _ready_to_exit = "--version";
```

pbehne commented on framework/src/base/MooseApp.C:
```diff
@@ -814,13 +814,13 @@ MooseApp::setupOptions()
   {
     _command_line->printUsage();
-    _ready_to_exit = true;
+    _ready_to_exit = "help";
```
```suggestion
    _ready_to_exit = "--help";
```

pbehne commented on framework/src/base/MooseApp.C:
```diff
@@ -843,7 +843,7 @@ MooseApp::setupOptions()
       Moose::out << "\n### START DUMP DATA ###\n"
                  << formatter.toString(tree.getRoot()) << "\n### END DUMP DATA ###" << std::endl;
-      _ready_to_exit = true;
+      _ready_to_exit = "dump";
```
```suggestion
      _ready_to_exit = "--dump";
```

pbehne commented on framework/src/base/MooseApp.C:
```diff
@@ -867,8 +867,7 @@ MooseApp::setupOptions()
     }
-
-    _ready_to_exit = true;
+    _ready_to_exit = "registry";
```
```suggestion
    _ready_to_exit = "--registry";
```

pbehne commented on framework/src/base/MooseApp.C:
```diff
@@ -913,7 +912,7 @@ MooseApp::setupOptions()
 
     Moose::out << "\n### END REGISTRY DATA ###\n";
-    _ready_to_exit = true;
+    _ready_to_exit = "registry_hit";
```
```suggestion
    _ready_to_exit = "--registry_hit";
```

pbehne commented on framework/src/base/MooseApp.C:
```diff
@@ -924,7 +923,7 @@ MooseApp::setupOptions()
     Moose::out << "%-START-SON-DEFINITION-%\n"
                << formatter.toString(tree.getRoot()) << "\n%-END-SON-DEFINITION-%\n";
-    _ready_to_exit = true;
+    _ready_to_exit = "definition";
```
```suggestion
    _ready_to_exit = "--definition";
```

pbehne commented on framework/src/base/MooseApp.C:
```diff
@@ -935,7 +934,7 @@ MooseApp::setupOptions()
     _builder.buildFullTree(search);
 
-    _ready_to_exit = true;
+    _ready_to_exit = "yaml";
```
```suggestion
    _ready_to_exit = "--yaml";
```

pbehne commented on framework/src/base/MooseApp.C:
```diff
@@ -947,7 +946,7 @@ MooseApp::setupOptions()
 
     Moose::out << "**START JSON DATA**\n" << tree.getRoot().dump(2) << "\n**END JSON DATA**\n";
-    _ready_to_exit = true;
+    _ready_to_exit = "json";
```
```suggestion
    _ready_to_exit = "--json";
```

pbehne commented on framework/src/base/MooseApp.C:
```diff
@@ -958,14 +957,14 @@ MooseApp::setupOptions()
       Moose::out << it.first << "\n";
     Moose::out << "**END SYNTAX DATA**\n" << std::endl;
-    _ready_to_exit = true;
+    _ready_to_exit = "syntax";
```
```suggestion
    _ready_to_exit = "--syntax";
```

pbehne commented on framework/src/base/MooseApp.C:
```diff
@@ -958,14 +957,14 @@ MooseApp::setupOptions()
 
     Moose::out << "MooseApp Type: " << type() << std::endl;
-    _ready_to_exit = true;
+    _ready_to_exit = "show_type";
```
```suggestion
    _ready_to_exit = "--show_type";
```

pbehne commented on framework/src/base/MooseApp.C:
```diff
@@ -1054,13 +1053,13 @@ MooseApp::setupOptions()
     moose_server.run();
 
-    _ready_to_exit = true;
+    _ready_to_exit = "language_server";
```
```suggestion
    _ready_to_exit = "--language_server";
```

pbehne commented on framework/src/base/MooseApp.C:
```diff
@@ -1112,17 +1111,19 @@ MooseApp::runInputFile()
   {
     // TODO: ask multiapps for their constructed objects
-    _ready_to_exit = true;
+    _ready_to_exit = "list_constructed_objects";
```
```suggestion
    _ready_to_exit = "--list-constructed-objects";
```

pbehne commented on modules/solid_mechanics/src/base/SolidMechanicsApp.C:
```diff
@@ -53,7 +53,7 @@ SolidMechanicsApp::runInputFile()
 
   if (getParam<bool>("parse_neml2_only"))
-    _ready_to_exit = true;
+    _ready_to_exit = "parse_neml2_only";
```
```suggestion
    _ready_to_exit = "--parse-neml2-only";
```

pbehne commented on framework/src/base/MooseApp.C:
```diff
@@ -1527,13 +1536,13 @@ MooseApp::run()
 
     Moose::out << docmsg << "\n";
-    _ready_to_exit = true;
+    _ready_to_exit = "show_docs";
```
```suggestion
    _ready_to_exit = "--docs";
```

pbehne commented on framework/src/base/MooseApp.C:
```diff
@@ -1139,6 +1140,14 @@ MooseApp::errorCheck()
 
   _builder.errorCheck(*_comm, warn, err);
 
+  if (!_executor.get() && !_executioner.get() && !_ready_to_exit.empty())
+    mooseError("Incompatible command line arguments provided. --check-input cannot be called with ",
+               _ready_to_exit,
+               ".");
+  else if (!_executor.get() && !_executioner.get())
+    mooseError("The Executor is being called without being initialized. This is likely caused by "
+               "incompatible command line arguments");
+
```
Why `!_executor.get()` AND `!_executioner.get()`? Why not `!_executor.get()` OR `!_executioner.get()`?

In either case, you can make this more efficient with
```
if (!_executor.get() && !_executioner.get())
{
  if (!_ready_to_exit.empty()
  {
    if (_check_input)
      mooseError("Incompatible command line arguments provided. --check-input cannot...");
    else
      mooseError("We should never get here!");
  }
  else
    mooseError("The Executor is being called...");
}
```
Not the added check for `_check_input`. Without it, hardcoding "--check-input" into the error message makes me uneasy. Do you know for sure that the code will only be executed if --check-input is provided?

DanielYankura replied:
The reason I did AND was because MooseApp::feProblem() segfaults only when both are uninitialized. For the params that I checked usually one would be uninitialized and then other initialized. I can definitely go ahead an make it more efficient though.

pbehne commented on test/tests/mesh/mesh_only/tests:
```diff
@@ -78,4 +78,13 @@
+    expect_err = "Incompatible command line arguments provided. --check-input cannot be called with mesh_only."
+    method = '!dbg'
+    issues = '#29788'
+    requirement = 'The system shall throw an error is the user provides a parameter that would result in a segfault when used with --mesh-only'
```
```suggestion
    requirement = 'The system shall throw an error if the user specifies incompatible command line arguments that would cause a segfault.'
```

pbehne commented on framework/include/base/MooseApp.h:
```diff
@@ -1251,7 +1251,7 @@ class MooseApp : public ConsoleStreamInterface,
   /// Indicates whether warnings or errors are displayed when overridden parameters are detected
   bool _error_overridden;
-  bool _ready_to_exit;
+  std::string _ready_to_exit;
```
Let's rename the variable `_ready_to_exit` to something more informative because we have changed it to a string. How about something like `_early_exit_reason`?

DanielYankura replied:
What about `_early_exit_param`, or would that be less clear as to what it does?

pbehne replied:
Sounds good to me!

DanielYankura committed:
various changes for PR

- changed _ready_to_exit to _early_exit_param
- _early_exit_param now matches what the user would input into the command line
- streamlined error check and removed uneeded function from MooseApp.h

pbehne reviewed (changes requested):
Looks good to me after these changes and after all the tests are passing.

shikhar413 reviewed (commented):
Took a quick look, some additional comments from me as well

pbehne commented on framework/src/base/MooseApp.C:
```diff
@@ -1139,6 +1140,24 @@ MooseApp::errorCheck()
+            _early_exit_param,
+            ".");
+      else
+        mooseError("We should never get here!");
```
Good point. I asked that the error be added to be explicit. However, I think that the logic could be changed to this:
```
  if (!_executor.get() && !_executioner.get())
  {
    if (!_early_exit_param.empty() && _check_input)
      mooseError(
          "Incompatible command line arguments provided. --check-input cannot be called with ",
          _early_exit_param,
          ".");
    mooseError("The Executor is being called without being initialized. This is likely "
                "caused by "
                "incompatible command line arguments");
  }
```
or this:
```
  if (!_executor.get() && !_executioner.get())
  {
    if (!_early_exit_param.empty())
      // We only get here if _check_input is true
      mooseError(
          "Incompatible command line arguments provided. --check-input cannot be called with ",
          _early_exit_param,
          ".");
    mooseError("The Executor is being called without being initialized. This is likely "
                "caused by "
                "incompatible command line arguments");
  }
```
or this:
```
  if (!_executor.get() && !_executioner.get())
  {
    if (!_early_exit_param.empty())
    {
      mooseAssert(_check_input, "Something went wrong, we should only get here if _check_input is true.");
      mooseError(
          "Incompatible command line arguments provided. --check-input cannot be called with ",
          _early_exit_param,
          ".");
    }
    mooseError("The Executor is being called without being initialized. This is likely "
                "caused by "
                "incompatible command line arguments");
  }
```

Just so long as we make it explicit that the expectation is for `_check_input` to be true, especially since it is hard-coded in the error message. This will help keep others from having to dig for confirmation like you just did.

DanielYankura replied:
I think the last one sounds the best. Having some error get thrown in case `_check_error` is false would be useful for developers who may later try improperly calling `errorCheck()`.

pbehne replied:
> I think the last one sounds the best. Having some error get thrown in case `_check_error` is false would be useful for developers who may later try improperly calling `errorCheck()`.

Also useful if the logic in errorCheck or execution is changed down the road.

DanielYankura committed:
More changes for PR

-Improved test requirement wording
-Added mooseAssert(_check_input) for future developers

pbehne commented:
The changes look good to me. I'm ready to approve once the tests are passing.

pbehne commented:
The following apps use the variable `_ready_to_exit`:
* Marmot
* Relap-7
* Sabertooth
* SAM

If we want to proceed with this PR as is, we'll have to rename the variable and make a PR for each app. I do not currently have access to Marmot and Sabertooth to do this. I have requested access to Marmot, but Sabertooth is not on NCRC and I haven't yet dug around to find where to get it.

Another option could be to leave `_ready_to_exit` as it was and define `_early_exit_param` as an entirely new variable in addition to `_ready_to_exit`. I'm happy to go either way.

DanielYankura commented:
Originally I didn't want to make a new variable because I felt it would be redundant. That being said I'm not sure how much `_ready_to_exit` is used in these other apps so it might be wise to leave it alone for now. I can put it back in for now and if we ever want to eliminate it later on I can always open a new PR

DanielYankura committed:
Added _ready_to_exit back in. MooseApp now has _ready_to_exit and _early_exit_param

DanielYankura commented:
By the way, I looked at the ones that are still failing. There are two tests that failed because of a timeout. I don't think this is caused by the changes I made.

pbehne commented:
> By the way, I looked at the ones that are still failing. There are two tests that failed because of a timeout. I don't think this is caused by the changes I made.

I typically invalidate those until they pass :)

moosebuild commented:
Job [OpenMPI](https://civet.inl.gov/job/2674870/) on 6534430 : invalidated by @DanielYankura

moosebuild commented:
Job [Modules parallel](https://civet.inl.gov/job/2674854/) on 6534430 : invalidated by @DanielYankura

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/2674873/) on 6534430 : invalidated by @shikhar413

shikhar413 commented:
Sorry invalidated the coverage test too early - please ignore

pbehne commented on framework/src/base/MooseApp.C:
```diff
@@ -1139,6 +1160,22 @@ MooseApp::errorCheck()
+          _early_exit_param,
+          ".");
+    }
+    mooseError("The Executor is being called without being initialized. This is likely "
```
Do we have a way to trigger this error? There is currently no coverage for it.

DanielYankura replied:
As far as I know there is not a way to trigger it right now. There are three cases where `errorCheck()` can get called; if `--check-input` is being used, if `__executor` is initialized, or if `_executioner` is initialized. I left it in there for future developers in case someone makes a change where `errorCheck()` can get called without one of those being true.

pbehne commented on framework/src/base/MooseApp.C:
```diff
@@ -1139,6 +1160,22 @@ MooseApp::errorCheck()
+          _early_exit_param,
+          ".");
+    }
+    mooseError("The Executor is being called without being initialized. This is likely "
```
```suggestion
    // We should never get here
    mooseError("The Executor is being called without being initialized. This is likely "
```

DanielYankura committed:
Added comment to clarify that the developer should never reach that error

pbehne reviewed (approved):
Thanks for the improvement!
